File: experiments/e0010_first_ensemble/backtest.py
"""Vectorized backtest harness for experiment e0010_first_ensemble.

Execution convention (see ``RATIONALE.md``): a strategy emits a target position
``signal_t`` in ``[-1, 1]`` using information available up to and including
``Close(t)``; the trade is executed at ``Open(t+1)``. Concretely the position
held over the interval ``[Open(t), Open(t+1)]`` is ``signal_{t-1}`` (i.e.
``signal.shift(1)``), and the per-bar return is the open-to-open return
``Open(t+1) / Open(t) - 1``. This is look-ahead free.

The EUR/USD series is treated as one continuous stream: bars are assumed evenly
spaced and gaps from inactive periods are ignored (used only for annualization).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

import numpy as np
import pandas as pd

_SECONDS_PER_YEAR = 365.25 * 24 * 3600


class Strategy(Protocol):
    """Minimal strategy interface.

    Implementations map a bars frame (``open, high, low, close, volume`` indexed
    by time) to a target-position series in ``[-1, 1]`` aligned to the same index.
    The value at ``t`` must depend only on data up to and including ``Close(t)``.
    """

    name: str

    def generate_signal(self, bars: pd.DataFrame) -> pd.Series: ...


@dataclass
class BacktestResult:
    """Outcome of a single backtest."""

    returns: pd.Series  # per-bar net strategy returns
    equity: pd.Series  # cumulative equity, starts near 1.0
    position: pd.Series  # effective position held during each bar
    metrics: dict[str, float]


def infer_periods_per_year(index: pd.Index) -> float:
    """Estimate bars-per-year from the (assumed even) bar spacing of the index."""
    if isinstance(index, pd.DatetimeIndex) and len(index) >= 2:
        median_delta = index.to_series().diff().dropna().median()
        seconds = median_delta.total_seconds()
        if seconds and seconds > 0:
            return _SECONDS_PER_YEAR / seconds
    raise ValueError("Cannot infer periods_per_year; pass it explicitly.")


def compute_metrics(
    returns: pd.Series,
    position: pd.Series,
    turnover: pd.Series,
    periods_per_year: float,
) -> dict[str, float]:
    """Core performance metrics for a per-bar net return stream."""
    n = len(returns)
    if n == 0:
        return {}

    total_return = float((1.0 + returns).prod() - 1.0)
    log_growth = float(np.log1p(returns).sum())
    exp_arg = min(log_growth * periods_per_year / n, 700.0)
    ann_return = float(np.expm1(exp_arg))

    std = float(returns.std(ddof=0))
    ann_vol = std * np.sqrt(periods_per_year)
    sharpe = float(returns.mean() / std * np.sqrt(periods_per_year)) if std > 0 else 0.0

    equity = (1.0 + returns).cumprod()
    drawdown = equity / equity.cummax() - 1.0
    max_drawdown = float(drawdown.min())

    active = position != 0.0
    hit_rate = float((returns[active] > 0).mean()) if bool(active.any()) else float("nan")
    ann_turnover = float(turnover.sum() * periods_per_year / n)

    return {
        "total_return": total_return,
        "ann_return": ann_return,
        "ann_vol": ann_vol,
        "sharpe": sharpe,
        "max_drawdown": max_drawdown,
        "hit_rate": hit_rate,
        "ann_turnover": ann_turnover,
        "n_bars": float(n),
    }


def backtest(
    bars: pd.DataFrame,
    signal: pd.Series,
    cost_per_turnover: float = 0.0,
    periods_per_year: float | None = None,
) -> BacktestResult:
    """Run a vectorized backtest of ``signal`` on ``bars``.

    ``cost_per_turnover`` is the cost (in return terms) charged per unit of
    position change, capturing spread + slippage + fees. For example, paying a
    half-spread of ~0.5 bp to flip from flat to fully long costs ``0.00005``.
    """
    signal = signal.reindex(bars.index).clip(-1.0, 1.0)

    bar_ret = bars["open"].shift(-1) / bars["open"] - 1.0  # Open(t) -> Open(t+1)
    position = signal.shift(1).fillna(0.0)  # decided at Close(t-1), live at Open(t)
    turnover = (position - position.shift(1).fillna(0.0)).abs()

    gross = position * bar_ret
    net = gross - turnover * cost_per_turnover

    valid = bar_ret.notna()
    net = net[valid]
    position = position[valid]
    turnover = turnover[valid]

    equity = (1.0 + net).cumprod()
    ppy = periods_per_year if periods_per_year is not None else infer_periods_per_year(bars.index)
    metrics = compute_metrics(net, position, turnover, ppy)

    return BacktestResult(returns=net, equity=equity, position=position, metrics=metrics)

File: experiments/e0020_fast_backtester_composer/src/backtest.rs
use crate::bar::Bar;
use crate::metrics::{self, Metrics};

/// Outcome of a single vectorized backtest run.
#[derive(Debug, Clone, PartialEq)]
pub struct BacktestResult {
    pub returns: Vec<f64>,
    pub equity: Vec<f64>,
    pub position: Vec<f64>,
    pub metrics: Metrics,
}

/// Run a vectorized backtest.
///
/// Execution convention (aligned with e0010): `signal[t]` uses information up to
/// and including `Close(t)`; the position held over `[Open(t), Open(t+1))` is
/// `signal[t-1]`. Per-bar return is open-to-open: `Open(t+1) / Open(t) - 1`.
pub fn backtest(bars: &[Bar], signal: &[f64], cost_per_turnover: f64) -> BacktestResult {
    let n = bars.len().min(signal.len());
    if n < 2 {
        return BacktestResult {
            returns: Vec::new(),
            equity: Vec::new(),
            position: Vec::new(),
            metrics: metrics::compute_metrics(&[], &[], &[], 60.0),
        };
    }

    let mut returns = Vec::with_capacity(n - 1);
    let mut equity = Vec::with_capacity(n - 1);
    let mut position_out = Vec::with_capacity(n - 1);
    let mut turnover_out = Vec::with_capacity(n - 1);

    let mut prev_position = 0.0;
    let mut cumulative_equity = 1.0;

    for i in 0..(n - 1) {
        let position = if i == 0 {
            0.0
        } else {
            signal[i - 1].clamp(-1.0, 1.0)
        };
        let bar_ret = bars[i + 1].open / bars[i].open - 1.0;
        let turnover = (position - prev_position).abs();
        let gross = position * bar_ret;
        let net = gross - turnover * cost_per_turnover;

        cumulative_equity *= 1.0 + net;
        returns.push(net);
        equity.push(cumulative_equity);
        position_out.push(position);
        turnover_out.push(turnover);
        prev_position = position;
    }

    let timestamps: Vec<_> = bars.iter().take(n).map(|b| b.timestamp).collect();
    let spacing = metrics::infer_bar_spacing_secs(&timestamps);
    let metrics = metrics::compute_metrics(&returns, &position_out, &turnover_out, spacing);

    BacktestResult {
        returns,
        equity,
        position: position_out,
        metrics,
    }
}

#[cfg(test)]
mod tests {
    use chrono::{FixedOffset, TimeZone};

    use super::*;

    fn bar(minute: u32, open: f64, close: f64) -> Bar {
        let tz = FixedOffset::east_opt(2 * 3600).unwrap();
        Bar {
            timestamp: tz.with_ymd_and_hms(2024, 1, 2, 0, minute, 0).unwrap(),
            open,
            high: open,
            low: open,
            close,
            volume: 1.0,
        }
    }

    #[test]
    fn flat_signal_has_zero_return() {
        let bars = vec![bar(0, 1.0, 1.0), bar(1, 1.1, 1.1), bar(2, 1.0, 1.0)];
        let signal = vec![0.0, 0.0, 0.0];
        let result = backtest(&bars, &signal, 0.0);
        assert!(result.returns.iter().all(|r| r.abs() < f64::EPSILON));
    }

    #[test]
    fn long_signal_tracks_open_to_open() {
        let bars = vec![bar(0, 1.0, 1.0), bar(1, 1.1, 1.1), bar(2, 1.21, 1.21)];
        let signal = vec![1.0, 1.0, 1.0];
        let result = backtest(&bars, &signal, 0.0);
        assert_eq!(result.returns.len(), 2);
        assert!(result.returns[0].abs() < f64::EPSILON);
        assert!((result.returns[1] - 0.1).abs() < 1e-12);
    }
}
